File: T-Bank/Algorithms-and-Data-Structures-2025/4sprint-SolvingProblemMethodsPart1/contest/C-Коровы_в_стойла/solution.py
import sys


def can_place(dist: int, k: int, nums: list[int]) -> bool:
    prev: int = 0
    for num in nums:
        if num >= prev:
            k -= 1 
            prev = num + dist
    return k <= 0


def solution() -> None:
    """
    Time Complexity O(NlogM)
    Memory Complexity O(1)
    """
    n, k = map(int, sys.stdin.readline().rstrip().split())
    nums: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
    left: int = 1
    right: int = nums[-1]
    answer: int = 0
    while left <= right:
        middle: int = left + ((right - left) >> 1)
        if can_place(middle, k, nums):
            answer = middle
            left = middle + 1 
        else: right = middle - 1 
    sys.stdout.write('{}\n'.format(answer))


if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/4sprint-SolvingProblemMethodsPart1/contest/C-Коровы_в_стойла/solution.cpp
#include <iostream>
#include <vector>



bool canPlace(int& dist, int k, std::vector<long long>& nums) {
    long long prev = 0;
    for (long long& num : nums) {
        if (num >= prev) {
            --k;
            prev = num + dist;
        }
    }
    return k <= 0;
}


void solution() {
    /*
    Time Complexity O(NlogM)
    Memory Complexity O(1)
    */
    int n, k;
    std::scanf("%d %d", &n, &k);
    std::vector<long long> nums (n, 0);
    for (int i = 0; i < n; ++i) std::scanf("%lld", &nums[i]);
    int left = 1;
    int right = nums[n - 1];
    int answer = 0;
    while (left <= right) {
        int middle = left + (right - left) / 2;
        if (canPlace(middle, k, nums)) {
            answer = middle;
            left = middle + 1;
        } else right = middle - 1;
    }
    std::printf("%d\n", answer);
}


int main() {
    solution();
}